lib/galaxy/tool_util/linters/tests.py

```diff
@@ -26,6 +26,12 @@ def lint_tsts(tool_xml, lint_ctx):
                 has_test = True
                 break
 
+        # check if expect_num_outputs is set if there are outputs with filters or discovered datasets
+        discover_datasets = tool_xml.findall("./outputs//discover_datasets")
+        filter = tool_xml.findall("./outputs//filter")
+        if len(discover_datasets) + len(filter) > 0 and "expect_num_outputs" not in test.attrib:
+            lint_ctx.warn("Test should specify 'expect_num_outputs' if outputs have filters or use discover_datasets")
```

Comment:
There was a discussion about this, maybe we should first do some consistent counting of collections and datasets and describe that.

Reply:
Adapted it to the latest state of https://github.com/galaxyproject/galaxy/pull/7894